### DISCUSSION TITLE
Poisson Nernst Planck Model

### DISCUSSION THREAD

**Discussion by nishant1785** [idaholab/moose#22144] [Q&A General] [answered]
Hello all,
I want to simulate ion transportation through an electrolyte for that I am using Poisson-Nernst-Planck model as shown below.

![image](https://user-images.githubusercontent.com/113843919/190957255-25919dc2-f499-423f-864d-29be5f390975.png)

I have used "soretDiffusion" module to formulate effect of electromigration effect (2nd term in Nernst-Planck equation). Initially linear potential profile is developed and it remains linear and also concentration of cation and anion doesn't change while residual keeps on increasing continuosly.

Plz help me i have to complete project by end of september.

Thank you very much,
Nishant Gaikwad

**Comment by GiudGiud** [comment]
Hello

I converted your post into a discussions as this is where we usually talk about these subjects
@tonkmr could you please help with this? Seems you are the author of Soret Diffusion in moose

Guillaume

**Reply by GiudGiud** [reply]
@nishant1785 you should make sure to follow the application development tutorial if you have not. It s very likely you will need to implement some new kernels to achieve this

**Comment by nishant1785** [comment]
Thank you very much Guillaume for your response I have followed application
development tutorial.  But I will recheck again.

Nishant Gaikwad-7972836355
M. Tech student
Materials Department
DIAT, DRDO,Girinagar
411025-Pune

**Comment by tonkmr** [comment]
I am happy to look at input files and discuss. You could also try to use the MatDiffusion kernel, where you use a coupled variable. That may be more flexible than SoretDiffusion.

**Comment by nishant1785** [comment]
Thank you Michael, instead of MatDiffusion I have used simple diffusion
module because I have to couple variables and I don't know how to do it in
MatDiffusion module. Now input files which I am attaching here are for
non-dimensional formulation of PNP model I mentioned earlier.

*FOR NON-DIMENSIONALISED CASE PERMITTIVITY,DIFFUSIVITY VANISHES FROM
EQUATION HENCE WE ARE ONLY LEFT WITH VARIABLES NO CONSTANTS OR 'EQUATING
ALL CONSTANTS EQUAL TO 1.*
I am doing this for symmetric binary electrolyte meaning* 'CATION VALENCY =
1, ANION VALENCY = -1'.*

I have also attached* boundary** conditions file.*

**Reply by GiudGiud** [reply]
hi @nishant1785 
the file did not paste from your email. You ll have to upload it from the github discussion page. 

If I were you I'd list each term in the equation and say how you implemented it.

**Reply by nishant1785** [reply]
@GiudGiud  is the provided information enough. Also @tonkmr i have used DIFFUSION KERNEL to code every part of PNP equation (i.e. for voltage/potential difference, diffusion of ions (separate equation for cation and anion) , electromigration effect)

**Comment by nishant1785** [comment]
[include.zip](https://github.com/idaholab/moose/files/9622137/include.zip)
[src.zip](https://github.com/idaholab/moose/files/9622140/src.zip)
[inputfile.zip](https://github.com/idaholab/moose/files/9622141/inputfile.zip)


sorry,Here are program files, in bounday condition image C₁ is cation concentration while C₂ is anion concentration. N is flux of ions which should be zero on every wall of a cube (kernel takes care of this boundary condition). More thoroughly my problem is to model ion flow in a cube by application of potential difference at left and right side of cube +1V and -1V.

Thank you.

**Reply by GiudGiud** [reply]
So in Diffusiona and c,
- you are missing the diffusion coefficient in both residual and jacobian
- you add a ca and a cc parameter, but arent you using these kernels in the same equation as the diffusion term uses (eg ca = u or cc = u)
Are you even using them? in the input file you are using Diffusion, not Diffusiona or Diffusionc

**Reply by nishant1785** [reply]
@GiudGiud  as I made equation dimensionless no need of diffusion coefficients as I mentioned before. Also "diffusiona and c" are for corresponding "electroa and c" because here I have to couple variables.

**Reply by GiudGiud** [reply]
Ok but diffusiona and c are not used in the input file. You have the regular Diffusion there. 

 if the equation is dimensionless don’t you need to have a factor on the time derivative ?

**Comment by nishant1785** [comment, marked as answer]
Earlier I used openFoam software in it program didn't run unless I gave
factor for time derivative but here it ran so i thought no need. Diffusiona
and c both of them are present in "electroa and c" Therefore I didn't put
it in input file.

Nishant Gaikwad-7972836355
M. Tech student
Materials Department
DIAT, DRDO,Girinagar
411025-Pune

**Reply by GiudGiud** [reply]
well if you want the right time evolution you will need to have the right coefficient in front the time derivatives. If you only care about the steady state, then it does not matter.

Ok I had not looked at the electroa and c terms. In those:
- I also dont see all the coefficients in front of the gradient_psi term. This is once again because the equation is dimensionless I guess, but the factor is not the same as for the diffusion term
- I dont think the Jacobian term is right. There should be an off-diagonal contribution given the involvement of psi, and the on-diagonal contribution should include grad_psi as well ?
If you want to post the derivation here we could discuss it

**Reply by nishant1785** [reply]
Thank you so much @GiudGiud  can i attach a paper in it on a particular page there is derivation of it and all. I don't have much idea about off-diagonal and on-diagonal terms. Please forgive me for 1 mistake I made "input file which I sent u is for uncoupled variables for diffusion terms in NERNST-PlANCK equation" Because I already tried coupling them & result was not converging anyway.

**Reply by GiudGiud** [reply]
Sure I can take a look

**Reply by nishant1785** [reply]
[jp402181e (2).pdf](https://github.com/idaholab/moose/files/9640503/jp402181e.2.pdf)

Take a look at page number 4 and 5.

**Reply by nishant1785** [reply]
![16641012776544120350291785446948](https://user-images.githubusercontent.com/113843919/192138743-f6aa487b-6632-4ed4-bd62-16cd812f06ea.jpg)

For electromigration term 👆 I tried soret wherein off-diagonal elements were computed.

**Reply by nishant1785** [reply]
Uploading VID_20220925_163421.mp4…

This linear profile should change & become more like exponentially decreasing but it doesn't change at all.

**Reply by GiudGiud** [reply]
hello

the video didnt load. and please avoid pictures of text, they are not searchable by others. 

The paper does not contain a derivation of the weak form, and of the Jacobian, so it's not helpful for debugging this. 

If you want to check your Jacobian you can use the jacobian_analyzer python script. 
https://mooseframework.inl.gov/help/development/analyze_jacobian.html

Guillaume

**Reply by nishant1785** [reply]
Okk I will do so for weak form see equation (22) in following paper. But this paper doesn't contain any information about jacobian. 
[lu2010.pdf](https://github.com/idaholab/moose/files/9641287/lu2010.pdf)

Here is the video... 

https://user-images.githubusercontent.com/113843919/192155308-2803dd3b-3d89-4e8c-a365-8e4fec039a1e.mp4

**Reply by GiudGiud** [reply]
Well exactly. The derivation of the Jacobian is the hard part, and it s likely where the mistake is. 

Would you consider using automatic differentiation?
It's a lot easier to write the kernels that way
see this: https://mooseframework.inl.gov/moose/modules/phase_field/FunctionMaterials/AutomaticDifferentiation.html

**Reply by GiudGiud** [reply]
This is probably a better documentation page for AD
https://mooseframework.inl.gov/automatic_differentiation/#!
I d really advise it to get the Jacobian right

**Comment by nishant1785** [comment]
Yeah will go through this too if this works.

Nishant Gaikwad-7972836355
M. Tech student
Materials Department
DIAT, DRDO,Girinagar
411025-Pune

**Comment by nishant1785** [comment]
Sure I will go through it.

Nishant Gaikwad-7972836355
M. Tech student
Materials Department
DIAT, DRDO,Girinagar
411025-Pune
